File: coco/pattern/sensec/app/crypto/dek_manager.py
"""
DEK lifecycle: generate, encrypt, decrypt using AES-256-GCM (FIPS 140-2 approved).

AAD (Additional Authenticated Data) binds ciphertext to the originating app_id.
This prevents a stolen ciphertext blob from being decrypted by a different app,
even if that app obtains the EDEK.
"""

from __future__ import annotations

import base64
import hashlib
import secrets
import struct
import uuid
from dataclasses import dataclass

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from app.crypto import iv_factory

# Expected byte lengths for AES-256-GCM parameters
IV_LENGTH = 12   # 96-bit nonce
TAG_LENGTH = 16  # 128-bit GCM authentication tag

# Token layout (binary, inside base64url wrapper):
#   1 byte  : format version (currently 0x01)
#  16 bytes : edek_id UUID (big-endian)
#  12 bytes : AES-GCM IV (nonce)
#  16 bytes : AES-GCM authentication tag
#   N bytes : ciphertext (variable)
#
# On-wire: "v1.<base64url(binary)>"
# The "v1." prefix lets future parsers detect the version before decoding.
_TOKEN_VERSION = 0x01
_TOKEN_PREFIX = "v1."
_TOKEN_FIXED_BYTES = 1 + 16 + IV_LENGTH + TAG_LENGTH   # 45 bytes


DEK_LENGTH_BYTES = 32   # 256-bit AES key
ALGORITHM = "AES-256-GCM"   # persisted per-record so future algorithm migrations stay decryptable


@dataclass(slots=True)
class EncryptResult:
    ciphertext: bytes
    iv: bytes
    tag: bytes          # GCM authentication tag (last 16 bytes of cryptography output)


def generate_dek() -> bytearray:
    """Return a fresh 256-bit DEK as a mutable bytearray so it can be zeroed."""
    raw = secrets.token_bytes(DEK_LENGTH_BYTES)
    return bytearray(raw)


def encrypt(plaintext: bytes, dek: bytearray, app_id: str) -> EncryptResult:
    """
    AES-256-GCM encrypt. The cryptography library appends the 16-byte GCM tag
    to the ciphertext; we split them for explicit storage.
    """
    iv = iv_factory.generate()
    aad = _make_aad(app_id)
    aesgcm = AESGCM(bytes(dek))
    combined = aesgcm.encrypt(iv, plaintext, aad)
    # cryptography returns ciphertext || tag
    ciphertext, tag = combined[:-16], combined[-16:]
    return EncryptResult(ciphertext=ciphertext, iv=iv, tag=tag)


def decrypt(ciphertext: bytes, tag: bytes, iv: bytes, dek: bytearray, app_id: str) -> bytes:
    """
    AES-256-GCM decrypt + tag verification. Raises cryptography.exceptions.InvalidTag
    on authentication failure (tampered ciphertext or wrong app_id).
    """
    aad = _make_aad(app_id)
    aesgcm = AESGCM(bytes(dek))
    combined = ciphertext + tag
    return aesgcm.decrypt(iv, combined, aad)


def zero_dek(dek: bytearray) -> None:
    """Overwrite DEK bytes in memory immediately after use."""
    for i in range(len(dek)):
        dek[i] = 0


def pack_token(edek_id: uuid.UUID, iv: bytes, tag: bytes, ciphertext: bytes) -> str:
    """
    Encode all decrypt inputs into one opaque token the client stores and echoes back.

    Binary layout: version(1) | edek_id(16) | iv(12) | tag(16) | ciphertext(N)
    On-wire:       "v1.<base64url(binary)>"
    """
    binary = bytes([_TOKEN_VERSION]) + edek_id.bytes + iv + tag + ciphertext
    return _TOKEN_PREFIX + base64.urlsafe_b64encode(binary).decode()


@dataclass(slots=True)
class UnpackedToken:
    edek_id: uuid.UUID
    iv: bytes
    tag: bytes
    ciphertext: bytes


def unpack_token(token: str) -> UnpackedToken:
    """
    Decode a ciphertext_token produced by pack_token().
    Raises ValueError with a descriptive message on any format error.
    """
    if not token.startswith(_TOKEN_PREFIX):
        raise ValueError(
            f"ciphertext_token has unrecognised format: expected prefix '{_TOKEN_PREFIX}'"
        )
    b64_part = token[len(_TOKEN_PREFIX):]
    try:
        binary = base64.urlsafe_b64decode(b64_part + "==")   # pad-safe
    except Exception:
        raise ValueError("ciphertext_token contains invalid base64url data")

    if len(binary) < _TOKEN_FIXED_BYTES:
        raise ValueError(
            f"ciphertext_token is too short: {len(binary)} bytes "
            f"(minimum {_TOKEN_FIXED_BYTES})"
        )

    version = binary[0]
    if version != _TOKEN_VERSION:
        raise ValueError(
            f"ciphertext_token uses unsupported version 0x{version:02x}; "
            f"this service supports 0x{_TOKEN_VERSION:02x}"
        )

    offset = 1
    edek_id = uuid.UUID(bytes=binary[offset:offset + 16]);  offset += 16
    iv      = binary[offset:offset + IV_LENGTH];             offset += IV_LENGTH
    tag     = binary[offset:offset + TAG_LENGTH];            offset += TAG_LENGTH
    ciphertext = binary[offset:]

    if not ciphertext:
        raise ValueError("ciphertext_token contains no ciphertext payload")

    return UnpackedToken(edek_id=edek_id, iv=iv, tag=tag, ciphertext=ciphertext)


def make_fingerprint(iv: bytes, tag: bytes) -> str:
    """
    First 8 bytes of SHA-256(iv || tag) as a 16-char hex string.
    Stored with the EDEK record so decrypt can detect element mix-ups
    before AES-GCM decryption even runs. Not a secret — it's a consistency
    check, not a MAC; the actual authentication is AES-GCM's tag.
    """
    return hashlib.sha256(iv + tag).hexdigest()[:16]


def _make_aad(app_id: str) -> bytes:
    return f"hsm-svc:app_id={app_id}".encode()

File: coco/pattern/springsec/hsm-core-service/src/main/java/com/hsm/core/crypto/DekManager.java
package com.hsm.core.crypto;

import org.bouncycastle.crypto.CryptoServicesRegistrar;
import org.bouncycastle.jcajce.provider.BouncyCastleFipsProvider;

import javax.crypto.AEADBadTagException;
import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.Base64;
import java.util.HexFormat;
import java.util.UUID;

/**
 * DEK lifecycle: generate, encrypt, decrypt using AES-256-GCM (FIPS 140-3 approved),
 * backed by the CMVP-validated Bouncy Castle FIPS Java API provider (cert #4943).
 *
 * <p>AAD (Additional Authenticated Data) binds ciphertext to the originating app_id.
 * This prevents a stolen ciphertext blob from being decrypted by a different app,
 * even if that app obtains the EDEK.
 *
 * <p>Token layout (binary, inside base64url wrapper):
 * <pre>
 *   1 byte  : format version (currently 0x01)
 *  16 bytes : edek_id UUID (big-endian)
 *  12 bytes : AES-GCM IV (nonce)
 *  16 bytes : AES-GCM authentication tag
 *   N bytes : ciphertext (variable)
 * </pre>
 * On-wire: {@code "v1.<base64url(binary)>"}. The "v1." prefix lets future
 * parsers detect the version before decoding.
 */
public final class DekManager {

    public static final int IV_LENGTH = 12;   // 96-bit nonce
    public static final int TAG_LENGTH = 16;  // 128-bit GCM authentication tag (bytes)
    private static final int TAG_LENGTH_BITS = TAG_LENGTH * 8;

    public static final int DEK_LENGTH_BYTES = 32; // 256-bit AES key
    public static final String ALGORITHM = "AES-256-GCM"; // persisted per-record so future algorithm migrations stay decryptable

    private static final String CIPHER_TRANSFORMATION = "AES/GCM/NoPadding";
    private static final String PROVIDER = BouncyCastleFipsProvider.PROVIDER_NAME;

    private static final byte TOKEN_VERSION = 0x01;
    private static final String TOKEN_PREFIX = "v1.";
    private static final int TOKEN_FIXED_BYTES = 1 + 16 + IV_LENGTH + TAG_LENGTH; // 45 bytes

    // FIPS-approved-mode DRBG-backed SecureRandom, rather than a plain java.security.SecureRandom.
    private static final SecureRandom RANDOM = CryptoServicesRegistrar.getSecureRandom();

    private DekManager() {
    }

    public record EncryptResult(byte[] ciphertext, byte[] iv, byte[] tag) {
    }

    public record UnpackedToken(UUID edekId, byte[] iv, byte[] tag, byte[] ciphertext) {
    }

    /** Return a fresh 256-bit DEK. Caller is responsible for zeroing it after use via {@link #zeroDek}. */
    public static byte[] generateDek() {
        byte[] raw = new byte[DEK_LENGTH_BYTES];
        RANDOM.nextBytes(raw);
        return raw;
    }

    /**
     * AES-256-GCM encrypt. The JCE cipher appends the 16-byte GCM tag to the
     * ciphertext; we split them for explicit storage, mirroring the Python
     * {@code cryptography} library's output shape.
     */
    public static EncryptResult encrypt(byte[] plaintext, byte[] dek, String appId) {
        byte[] iv = IvFactory.generate();
        byte[] aad = makeAad(appId);
        try {
            Cipher cipher = Cipher.getInstance(CIPHER_TRANSFORMATION, PROVIDER);
            cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(dek, "AES"), new GCMParameterSpec(TAG_LENGTH_BITS, iv));
            cipher.updateAAD(aad);
            byte[] combined = cipher.doFinal(plaintext);
            int ctLen = combined.length - TAG_LENGTH;
            byte[] ciphertext = Arrays.copyOfRange(combined, 0, ctLen);
            byte[] tag = Arrays.copyOfRange(combined, ctLen, combined.length);
            return new EncryptResult(ciphertext, iv, tag);
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("AES-GCM encryption failed", e);
        }
    }

    /**
     * AES-256-GCM decrypt + tag verification. Throws {@link AEADBadTagException}
     * on authentication failure (tampered ciphertext or wrong app_id) — mirrors
     * Python's {@code cryptography.exceptions.InvalidTag}.
     */
    public static byte[] decrypt(byte[] ciphertext, byte[] tag, byte[] iv, byte[] dek, String appId) throws AEADBadTagException {
        byte[] aad = makeAad(appId);
        byte[] combined = new byte[ciphertext.length + tag.length];
        System.arraycopy(ciphertext, 0, combined, 0, ciphertext.length);
        System.arraycopy(tag, 0, combined, ciphertext.length, tag.length);
        try {
            Cipher cipher = Cipher.getInstance(CIPHER_TRANSFORMATION, PROVIDER);
            cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(dek, "AES"), new GCMParameterSpec(TAG_LENGTH_BITS, iv));
            cipher.updateAAD(aad);
            return cipher.doFinal(combined);
        } catch (AEADBadTagException e) {
            throw e;
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("AES-GCM decryption failed", e);
        }
    }

    /** Overwrite DEK bytes in memory immediately after use. */
    public static void zeroDek(byte[] dek) {
        Arrays.fill(dek, (byte) 0);
    }

    /**
     * Encode all decrypt inputs into one opaque token the client stores and echoes back.
     * Binary layout: version(1) | edek_id(16) | iv(12) | tag(16) | ciphertext(N).
     * On-wire: "v1.&lt;base64url(binary)&gt;"
     */
    public static String packToken(UUID edekId, byte[] iv, byte[] tag, byte[] ciphertext) {
        ByteBuffer buf = ByteBuffer.allocate(TOKEN_FIXED_BYTES + ciphertext.length);
        buf.put(TOKEN_VERSION);
        buf.putLong(edekId.getMostSignificantBits());
        buf.putLong(edekId.getLeastSignificantBits());
        buf.put(iv);
        buf.put(tag);
        buf.put(ciphertext);
        return TOKEN_PREFIX + Base64.getUrlEncoder().encodeToString(buf.array());
    }

    /**
     * Decode a ciphertext_token produced by {@link #packToken}.
     * Throws {@link IllegalArgumentException} with a descriptive message on any format error.
     */
    public static UnpackedToken unpackToken(String token) {
        if (!token.startsWith(TOKEN_PREFIX)) {
            throw new IllegalArgumentException(
                    "ciphertext_token has unrecognised format: expected prefix '" + TOKEN_PREFIX + "'");
        }
        String b64Part = token.substring(TOKEN_PREFIX.length());
        byte[] binary;
        try {
            binary = Base64.getUrlDecoder().decode(padBase64(b64Part));
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("ciphertext_token contains invalid base64url data");
        }

        if (binary.length < TOKEN_FIXED_BYTES) {
            throw new IllegalArgumentException(
                    "ciphertext_token is too short: " + binary.length + " bytes (minimum " + TOKEN_FIXED_BYTES + ")");
        }

        byte version = binary[0];
        if (version != TOKEN_VERSION) {
            throw new IllegalArgumentException(String.format(
                    "ciphertext_token uses unsupported version 0x%02x; this service supports 0x%02x",
                    version, TOKEN_VERSION));
        }

        ByteBuffer buf = ByteBuffer.wrap(binary);
        buf.get(); // version, already validated
        long msb = buf.getLong();
        long lsb = buf.getLong();
        UUID edekId = new UUID(msb, lsb);
        byte[] iv = new byte[IV_LENGTH];
        buf.get(iv);
        byte[] tag = new byte[TAG_LENGTH];
        buf.get(tag);
        byte[] ciphertext = new byte[buf.remaining()];
        buf.get(ciphertext);

        if (ciphertext.length == 0) {
            throw new IllegalArgumentException("ciphertext_token contains no ciphertext payload");
        }

        return new UnpackedToken(edekId, iv, tag, ciphertext);
    }

    /**
     * First 8 bytes of SHA-256(iv || tag) as a 16-char hex string.
     * Stored with the EDEK record so decrypt can detect element mix-ups
     * before AES-GCM decryption even runs. Not a secret — it's a consistency
     * check, not a MAC; the actual authentication is AES-GCM's tag.
     */
    public static String makeFingerprint(byte[] iv, byte[] tag) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            digest.update(iv);
            digest.update(tag);
            byte[] hash = digest.digest();
            return HexFormat.of().formatHex(hash, 0, 8);
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException(e);
        }
    }

    private static byte[] makeAad(String appId) {
        return ("hsm-svc:app_id=" + appId).getBytes(StandardCharsets.UTF_8);
    }

    private static String padBase64(String s) {
        int mod = s.length() % 4;
        return mod == 0 ? s : s + "=".repeat(4 - mod);
    }
}
